File: backend/src/main/java/com/backend/api/payment/service/PaymentService.java
package com.backend.api.payment.service;


import com.backend.api.payment.dto.request.PaymentRequest;
import com.backend.api.payment.dto.response.PaymentConfirmResponse;
import com.backend.api.payment.dto.response.PaymentResponse;
import com.backend.domain.payment.entity.Payment;
import com.backend.domain.payment.entity.PaymentStatus;
import com.backend.domain.payment.repository.PaymentRepository;
import com.backend.domain.user.entity.User;
import com.backend.global.Rq.Rq;
import com.backend.global.exception.ErrorCode;
import com.backend.global.exception.ErrorException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatusCode;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.LocalDateTime;

//2차 때 사용 x
@Service
@RequiredArgsConstructor
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final WebClient webClient;
    private final Rq rq;


    //결제 승인
    @Transactional
    public PaymentResponse confirmPayment(PaymentRequest request){
        User user = rq.getUser();
        PaymentConfirmResponse response = webClient.post()
                .uri("/confirm") // 결제 승인 API 호출
                .bodyValue(request)
                .retrieve()
                .onStatus(HttpStatusCode::isError, clientResponse ->
                        clientResponse.bodyToMono(String.class)
                                .flatMap(errorBody ->
                                        Mono.error(new ErrorException(errorBody, ErrorCode.PAYMENT_APPROVE_FAILED))
                                )
                )
                .bodyToMono(PaymentConfirmResponse.class)
                .block();


        Payment payment = Payment.builder()
                .orderId(response.orderId())
                .paymentKey(response.paymentKey())
                .orderName(response.orderName())
                .method(response.method())
                .totalAmount(response.totalAmount())
                .status(PaymentStatus.DONE)
                .approvedAt(LocalDateTime.now())
                .user(user)
                .build();
        paymentRepository.save(payment);

        return PaymentResponse.from(payment);

    }

    @Transactional(readOnly = true)
    public PaymentResponse geyPaymentByKey(String paymentKey) {
        Payment payment = paymentRepository.findByPaymentKey(paymentKey)
                .orElseThrow(() -> new ErrorException(ErrorCode.PAYMENT_NOT_FOUND));

        return PaymentResponse.from(payment);
    }

    @Transactional(readOnly = true)
    public PaymentResponse getPaymentByOrderId(String orderId) {
        Payment payment = paymentRepository.findByOrderId(orderId)
                .orElseThrow(() -> new ErrorException(ErrorCode.PAYMENT_NOT_FOUND));

        return PaymentResponse.from(payment);
    }

    @Transactional
    public void savePayment(Payment payment) {
        paymentRepository.save(payment);
    }

}

File: backend/src/main/java/com/backend/api/payment/service/PaymentService.kt
package com.backend.api.payment.service

import com.backend.api.payment.dto.request.PaymentRequest
import com.backend.api.payment.dto.response.PaymentConfirmResponse
import com.backend.api.payment.dto.response.PaymentResponse
import com.backend.api.payment.dto.response.PaymentResponse.Companion.from
import com.backend.domain.payment.entity.Payment
import com.backend.domain.payment.entity.PaymentStatus
import com.backend.domain.payment.repository.PaymentRepository
import com.backend.global.Rq.Rq
import com.backend.global.exception.ErrorCode
import com.backend.global.exception.ErrorException
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono
import java.time.LocalDateTime

@Service
class PaymentService(
    private val paymentRepository: PaymentRepository,
    private val webClient: WebClient,
    private val rq: Rq
) {

    //결제 승인
    @Transactional
    fun confirmPayment(request: PaymentRequest): PaymentResponse {
        val user = rq.getUser()
        val response = webClient.post()
            .uri("/confirm") // 결제 승인 API 호출
            .bodyValue(request)
            .retrieve()
            .onStatus({ it.isError }) { clientResponse ->
                clientResponse.bodyToMono(String::class.java)
                    .flatMap { errorBody ->
                        Mono.error(ErrorException(errorBody, ErrorCode.PAYMENT_APPROVE_FAILED))
                    }
            }
            .bodyToMono(PaymentConfirmResponse::class.java)
            .block()
            ?: throw ErrorException(ErrorCode.PAYMENT_APPROVE_FAILED)

        val payment = Payment(
            orderId = response.orderId,
            paymentKey = response.paymentKey,
            orderName = response.orderName,
            totalAmount = response.totalAmount,
            method = response.method,
            status = PaymentStatus.DONE,
            approvedAt = LocalDateTime.now(),
            user = user
        )

        paymentRepository.save(payment)

        return from(payment)
    }

    @Transactional(readOnly = true)
    fun geyPaymentByKey(paymentKey: String): PaymentResponse {
        val payment= paymentRepository.findByPaymentKey(paymentKey)
            ?: throw ErrorException(ErrorCode.PAYMENT_NOT_FOUND)

        return from(payment)
    }

    @Transactional(readOnly = true)
    fun getPaymentByOrderId(orderId: String): PaymentResponse {
        val payment = paymentRepository.findByOrderId(orderId)
            ?: throw ErrorException(ErrorCode.PAYMENT_NOT_FOUND)

        return from(payment)
    }

    @Transactional
    fun savePayment(payment: Payment) {
        paymentRepository.save(payment)
    }
}
